processor/src/test/java/org/mapstruct/ap/test/context/ContextParameterTest.java

```diff
@@ -0,0 +1,117 @@
+/**
+ *  Copyright 2012-2016 Gunnar Morling (http://www.gunnarmorling.de/)
+ *  and/or other contributors as indicated by the @authors tag. See the
+ *  copyright.txt file in the distribution for a full listing of all
+ *  contributors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.mapstruct.ap.test.context;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mapstruct.BeforeMapping;
+import org.mapstruct.Context;
+import org.mapstruct.ObjectFactory;
+import org.mapstruct.ap.test.context.Node.Attribute;
+import org.mapstruct.ap.test.context.NodeDto.AttributeDto;
+import org.mapstruct.ap.testutil.IssueKey;
+import org.mapstruct.ap.testutil.WithClasses;
+import org.mapstruct.ap.testutil.runner.AnnotationProcessorTestRunner;
+
+/**
+ * Tests the usage of the {@link Context} annotation in the following situations:
+ * <ul>
+ * <li>passing the parameter to property mapping methods (create and update)
+ * <li>passing the parameter to forged iterable methods
+ * <li>passing the parameter to forged bean mapping methods
+ * <li>passing the parameter to factory methods (with and without {@link ObjectFactory})
+ * <li>passing the parameter to lifecycle methods (in this case, {@link BeforeMapping}
+ * <li>passing multiple parameters, with varied order of context params and mapping source params
+ * </ul>
+ *
+ * @author Andreas Gudian
+ */
+@IssueKey("975")
+@WithClasses({
+    Node.class,
+    NodeDto.class,
+    NodeMapperWithContext.class,
+    AutomappingNodeMapperWithContext.class,
+    CycleContext.class,
+    FactoryContext.class,
+    CycleContextLifecycleMethods.class })
+@RunWith(AnnotationProcessorTestRunner.class)
+public class ContextParameterTest {
+
+    private static final int MATIC_NUMBER_OFFSET = 10;
```

Comment:
Should this be `MAGIC_NUMBER_OFFSET` instead of `MATIC_NUMBER_OFFSET`? There is also one variable name in the `FactoryContext` named `attributeMaticNumberOffset`. It is not a big deal, just something that confused me while having a look 😄